test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +523,668 @@ func TestRSSelectorImmutability(t *testing.T) {
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := c.Core().Pods(ns.Name).List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 1)
+	}
+
+	pod := &pods.Items[0]
+	newLabelMap := map[string]string{"new-foo": "new-bar"}
+	podClient := c.Core().Pods(ns.Name)
+
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		pod.Labels = newLabelMap
+		pod, err = c.Core().Pods(ns.Name).Update(pod)
+		if err != nil {
+			return fmt.Errorf("failed to update pod %s: %v", pod.Name, err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if metav1.GetControllerOf(pod) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Adoption: RS should add ControllerRef to a pod when the pod's labels change to match its labels
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Revert to original labels so that the RS can adopt the pod again
+		pod.Labels = labelMap
+		pod, err = c.Core().Pods(ns.Name).Update(pod)
+		if err != nil {
+			return fmt.Errorf("failed to update pod %s: %v", pod.Name, err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if metav1.GetControllerOf(pod) != nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// We have to get the RS again to fetch its newest UID
+	// the UID may change as time passes
```

Comment:
It's not that the UID changes over time. It's just that the UID is generated on the server side during the call to `Create()`, and you have not stored the result from that.

You won't need to Get the RS here if you modify `createRSsPods()` to return the lists of RSes and Pods that come back from `Create()`. Then after you call `rss, _ := createRSsPods()` above, set `rs = rss[0]`.

Reply:
Sure! I will fetch the UID in this way. Btw, what happens if i explicitly set UID for RS before calling `createRSsPods()`? From debugging, it seemed that the explicitly set RS will be replaced. Correct me if i am wrong.

Reply:
That's right. Create() will ignore any UID you set. The server generates its own UID and overwrites it.

Reply:
Thanks for confirming it!